flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FSDataBufferedInputStream.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.filesystem;
+
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * FSDataInputStream with Buffer
+ */
+public class FSDataBufferedInputStream extends FSDataInputStream {
+
+	private static final int DEFAULT_BUFFER_SIZE = 8192;
+
+	protected volatile byte[] buf;
+
+	// read offset of buf
+	private int pos;
+
+	// availed count of buf
+	private int count;
+
+	private final FSDataInputStream inputStream;
+
+	private volatile boolean closed;
+
+	public FSDataBufferedInputStream(FSDataInputStream inputStream) {
+		this(inputStream, DEFAULT_BUFFER_SIZE);
+	}
+
+	public FSDataBufferedInputStream(
+		FSDataInputStream inputStream,
+		int bufferSize) {
+		this.inputStream = inputStream;
+
+		Preconditions.checkState(bufferSize > 0,
+			new IllegalArgumentException("bufferSize must > 0"));
```

Comment:
We should pass error message instead of `IllegalArgumentException` here.

Reply:
done